libcontainer/container_linux.go

```diff
@@ -341,6 +341,13 @@ func (c *Container) start(process *Process) (retErr error) {
 	if err := utils.CloseExecFrom(3); err != nil {
 		return fmt.Errorf("unable to mark non-stdio fds as cloexec: %w", err)
 	}
+
+	// Tell the kernel that runc wants to reap orphaned children of the
+	// `runc init` process.
+	if err := unix.Prctl(unix.PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0); err != nil {
```

Comment:
I disagree with the entire approach, but if you want to do this you need to have add a `waitid(-1)` thread somewhere to make sure you actually reap the zombies.

Reply:
It has been waited the third init process before.

Reply:
https://github.com/opencontainers/runc/blob/09eacaa10b83cb8ff1d7d9a42df857b6c475f0fa/libcontainer/process_linux.go#L406-L411
https://github.com/opencontainers/runc/blob/09eacaa10b83cb8ff1d7d9a42df857b6c475f0fa/libcontainer/process_linux.go#L279
https://github.com/opencontainers/runc/blob/09eacaa10b83cb8ff1d7d9a42df857b6c475f0fa/libcontainer/process_linux.go#L328
https://github.com/opencontainers/runc/blob/09eacaa10b83cb8ff1d7d9a42df857b6c475f0fa/libcontainer/process_linux.go#L801

Reply:
`p.wait` doesn't do a `waitid(-1)` AFAIK, it only waits for the specific child process?

Reply:
You are right, we need `waitpid(-1)` here, we should not only wait `runc init`, but also processes created by `runc init`, for example, `newuidmap`, `newgidmap` etc.

Reply:
> but also processes created by runc init, for example, newuidmap, newgidmap etc.

`try_mapping_tool` already does a `waitpid`. Now that I think about it, we should do the wait in `nsexec.c` and not use `PR_SET_CHILD_SUBREAPER` (see my below comment). Setting `PR_SET_CHILD_SUBREAPER` implicitly is just asking for trouble.

Reply:
> `try_mapping_tool` already does a `waitpid`.

Yes, but if stage-0 process suddenly dead, I think runc should reap all these process.
As we all know, PR_SET_CHILD_SUBREAPER means that runc has the ability to reap grand child processes, but if the child process has not exited, runc will not reap it's grand child process.

I think the only problem is that we should tell the libct/nsexec users to add unix.Prctl(unix.PR_SET_CHILD_SUBREAPER, 1, 0, 0, 0) in their system like in runc.

Reply:
Ah. Right, stage-2 is not going to be a child of the original runc so we can't wait for it there... That's unfortunate...

The reason I don't like adding `PR_SET_CHILD_SUBREAPER` is that it's a process-wide setting so other users of `nsenter` (for example, as a server spawning containers) that spawn other processes will now have to deal with subreaper semantics (and depending on their design they might really not want that).

(And, given that this only fixes the surface-level issue with glibc it feels overkill to do `PR_SET_CHILD_SUBREAPER`... I suspect we will eventually need to work around other issues and will end up doing an `execve` so we might as well rip that band-aid off now...)

Reply:
Further think, except the three `runc init` process, what other processes should be reaped by runc process?
So there are two questions need to consider:
1. Do we really need `waitid(-1)`? Although add `waitid(-1)` has no hurt;
2. Do we need to open a new go routine to do reap before container terminated?